src/runtime/hexagon/hexagon_buffer_manager.h

```diff
@@ -85,6 +86,18 @@ class HexagonBufferManager {
     return hexagon_buffer_map_.empty();
   }
 
+  //! \brief Returns a vector of currently allocated pointers, owned by the manager.
+  // Note - this should only be used by the device API to keep track of what
+  // was in the manager when HexagonDeviceAPI::ReleaseResources is called.
+  std::vector<void*> current_allocations() {
```

Comment:
I apologize for the post-merge feedback.  But ... having this method seems like a break in the abstraction of the `HexagonBufferManager` class, to me.  If we have to do this it feels like something else is amiss.

Reply:
Thanks for the review.

I agree, this isn't ideal.  However, I wanted to make sure that I don't give a "free pass" to all allocations that come after ReleaseResources.  The scenario I was trying to avoid was:

User allocates buffer B
Session ends, ReleaseResources is called and B is freed
User frees buffer B as objects are being torn down, which should not throw as it is somewhat expected
User frees buffer C, which SHOULD throw, as it was never allocated in the first place

I added this so that I can check if there's no runtime_hexbuffs, then I make sure that what is being attempted to be freed is something that we owned before ReleaseResources was called.  In which case, that Free is a no-op.  I also want to detect spurious calls to Free that wouldn't have been legit before ReleaseResources was called.

I'm happy to discuss more, and brainstorm other ways to accomplish this.

Reply:
As an alternative, what if the `HexagonBufferManager` allows three states for buffers instead of two.

* Valid to use, valid to free (Initial state after allocating memory)
* Invalid to use, valid to free (State after closing a session)
* Invalid to use, invalid to free (State of unknown pointer)

That way, instead of discarding the `HexagonBufferManager runtime_hexbuffs` altogether when the session closes, it instead moves all keys from the internal `std::unordered_map<void*, std::unique_ptr<HexagonBuffer>> hexagon_buffer_map_;` to an internal `std::unordered_set<void*>`.  This avoids exposing any of the buffers to the outside, but does allow the `HexagonBufferManager` to identify a previously freed pointer.

Reply:
Thanks @Lunderberg, I think your suggestion would satisfy my concern.  What do you think @janetsc?  This would allow us to make valid frees a `nop` and to catch invalid frees as well.  If I am understanding correctly `runtime_hexbuffs` would be instantiated by the Hexagon Device API in the same way that we instantiate the `rpc_hexbuffs` ... `HexagonBufferManager runtime_hexbuffs;`?  If that is the case, we could potentially have just one HeagonBufferManager that implements the API and internal state that @Lunderberg suggests.

Reply:
There's also another state not captured - the buffer manager is in an invalid state to allocate more memory.

There needs to be two buffer managers, because on ReleaseResources how do we know which buffers need to be freed?

I'm thinking that a better model is to enforce that all frees should take place before ReleaseResources.  That's cleaner.  I will investigate which ones are outstanding.  All objects should be cleanly torn down by the time the session ends.  I will file a task to investigate.